### PR TITLE
[SPARK-45207][SQL][CONNECT] Implement Error Enrichment for Scala Client

### PR DESCRIPTION
**Pull request by heyihong**
### What changes were proposed in this pull request?

-  Implemented the reconstruction of the complete exception (un-truncated error messages, cause exceptions, server-side stacktrace) based on the responses of FetchErrorDetails RPC.

### Why are the changes needed?

- Cause exceptions play an important role in the current control flow, such as in StreamingQueryException. They are also valuable for debugging.
- Un-truncated error message is useful for debugging
- Providing server-side stack traces aids in effectively diagnosing server-related issues.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

- `build/sbt "connect-client-jvm/testOnly *ClientE2ETestSuite"`
- `build/sbt "connect-client-jvm/testOnly *ClientStreamingQuerySuite"`

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/CustomSparkConnectBlockingStub.scala [hunk 1]
```diff
@@ -27,11 +27,17 @@ private[connect] class CustomSparkConnectBlockingStub(
     retryPolicy: GrpcRetryHandler.RetryPolicy) {
 
   private val stub = SparkConnectServiceGrpc.newBlockingStub(channel)
+
   private val retryHandler = new GrpcRetryHandler(retryPolicy)
 
+  // Constructing GrpcExceptionConverter with a GRPC stub for fetching error details from server.
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/CustomSparkConnectBlockingStub.scala:33]: I'd remove the `Constructing` bit, because that this is pretty clear if you do `new`,,,,

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/ClientStreamingQuerySuite.scala [hunk 2]
```diff
@@ -175,6 +175,37 @@ class ClientStreamingQuerySuite extends QueryTest with SQLHelper with Logging {
...
+      .outputMode("append")
+      .format("console")
+      .start()
+
+    val exception = intercept[SparkException] {
+      query.awaitTermination()
+    }
+
+    assert(
+      exception.getCause.getCause.getCause.getMessage
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/ClientStreamingQuerySuite.scala:204]: Remind me again what is the chain here? And why are we not checking this?

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 3]
```diff
@@ -24,49 +24,135 @@ import scala.reflect.ClassTag
...
+      ex: StatusRuntimeException,
+      sessionId: String,
+      userContext: UserContext): Throwable = {
+    val status = StatusProto.fromThrowable(ex)
+
+    val errorInfoOpt = status.getDetailsList.asScala
+      .find(_.is(classOf[ErrorInfo]))
+      .map(_.unpack(classOf[ErrorInfo]))
+
+    if (errorInfoOpt.isDefined) {
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:141]: The code below implements the policy you described in the documentation. Please add comments for every step of the way.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 4]
```diff
@@ -93,33 +179,65 @@ private[client] object GrpcExceptionConverter extends JsonUtils {
...
+    val error = errors(errorIdx)
+
+    val classHierarchy = error.getErrorTypeHierarchyList.asScala
+
+    val constructor =
+      classHierarchy
+        .flatMap(errorFactory.get)
+        .headOption
+        .getOrElse((message: String, cause: Option[Throwable]) =>
+          new SparkException(
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:203]: How about we just do `s"${classHierarchy.head}: ${message}"`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -2882,8 +2882,7 @@ object SQLConf {
         "level settings.")
       .version("3.0.0")
       .booleanConf
-      // show full stacktrace in tests but hide in production by default.
-      .createWithDefault(Utils.isTesting)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2886]: Why is this needed?

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 6]
```diff
@@ -24,49 +24,135 @@ import scala.reflect.ClassTag
...
-        convert {
+        convert(sessionId, userContext) {
           iter.close()
         }
       }
     }
   }
 
+  /**
+   * fetchEnrichedError fetches enriched errors with full exception message and optionally
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:92]: remove the method name from the doc?

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala (hunk 4 above)
Comment (heyihong) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:203]: This is from https://github.com/apache/spark/pull/42377#discussion_r1329224205. But this one seems to be simpler

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (heyihong) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2886]: The size of the ErrorInfo is about to exceed the header limit even in tests. So disable this by default.

Comment (heyihong) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2886]: It is disabled by default in python connect tests already: https://github.com/apache/spark/blob/c89221b02bb3000f707a31322e6d40b561e527bd/python/pyspark/testing/connectutils.py#L173

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/ClientStreamingQuerySuite.scala (hunk 2 above)
Comment (heyihong) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/ClientStreamingQuerySuite.scala:204]: https://github.com/apache/spark/pull/42377#discussion_r1323060159

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2897]: Can you rebase and show me it is not an issue anymore.

Comment (heyihong) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Here is the fix: https://github.com/apache/spark/pull/43017

**Comment by heyihong** [comment]
@juliuszsompolski Please take a look

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 7]
```diff
@@ -24,49 +24,140 @@ import scala.reflect.ClassTag
...
+      return None
+    }
+
+    try {
+      val errorDetailsResponse = grpcStub.fetchErrorDetails(
+        FetchErrorDetailsRequest
+          .newBuilder()
+          .setSessionId(sessionId)
+          .setErrorId(errorId)
+          .setUserContext(UserContext.newBuilder().setUserId(userContext.getUserId).build())
```
Comment (juliuszsompolski) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:111]: why build new UserContext instead of reusing `userContext`? Why does only UserId field matter?

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 8]
```diff
@@ -93,33 +184,63 @@ private[client] object GrpcExceptionConverter extends JsonUtils {
...
-    errorInfoToThrowable(errorInfoOpt.get.unpack(classOf[ErrorInfo]), status.getMessage)
-      .getOrElse(fallbackEx)
+  /**
+   * errorInfoToThrowable reconstructs the exception based on the error classes hierarchy and the
+   * truncated error message.
+   */
+  private def errorInfoToThrowable(info: ErrorInfo, message: String): Throwable = {
+    implicit val formats = DefaultFormats
+    val classes =
+      JsonMethods.parse(info.getMetadataOrDefault("classes", "[]")).extract[Array[String]]
```
[the remark this answers is not in the document]
Comment (heyihong) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:235]: When the header size exceeds limit, the server will return errors in responses instead of truncating data silently. If the json is malformed, it may be better to fail early

**Review by juliuszsompolski — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.
